packages/gatsby/src/schema/node-model.js

```diff
@@ -0,0 +1,207 @@
+// @flow
+
+const _ = require(`lodash`)
+const {
+  isAbstractType,
+  GraphQLOutputType,
+  GraphQLUnionType,
+} = require(`graphql`)
+const invariant = require(`invariant`)
+
+type IDOrNode = string | { id: string }
+type TypeOrTypeName = string | GraphQLOutputType
+
+interface ConnectionArgs {
+  path: string;
+  connectionType?: string;
+}
+
+export interface NodeModel {
+  getNodeById(
+    { id: IDOrNode, type?: TypeOrTypeName },
+    ConnectionArgs
+  ): any | null;
+  getNodesByIds(
+    { ids: Array<IDOrNode>, type?: TypeOrTypeName },
+    ConnectionArgs
+  ): Array<any>;
+  getAllNodes({ type?: TypeOrTypeName }, ConnectionArgs): Array<any>;
+  runQuery(args: any): Promise<any>;
+  getTypes(): Array<string>;
+}
+
+class LocalNodeModel {
+  constructor({ schema, nodeStore, createPageDependency }) {
+    this.schema = schema
+    this.nodeStore = nodeStore
+    this.createPageDependency = createPageDependency
+  }
+
+  getNodeById(args, pageDependencies) {
+    const { id, type } = args || {}
+
+    const node = getNodeById(this.nodeStore, id)
+
+    let result
+    if (!node) {
+      result = null
+    } else if (!type) {
+      result = node
+    } else {
+      const nodeTypeNames = toNodeTypeNames(this.schema, type)
+      result = nodeTypeNames.includes(node.internal.type) ? node : null
+    }
+
+    return trackPageDependencies(
+      result,
+      pageDependencies,
+      this.createPageDependency
+    )
+  }
+
+  getNodesByIds(args, pageDependencies) {
+    const { ids, type } = args || {}
+
+    const nodes = Array.isArray(ids)
+      ? ids.map(id => getNodeById(this.nodeStore, id)).filter(Boolean)
+      : []
+
+    let result
+    if (!nodes.length || !type) {
+      result = nodes
+    } else {
+      const nodeTypeNames = toNodeTypeNames(this.schema, type)
+      result = nodes.filter(node => nodeTypeNames.includes(node.internal.type))
+    }
+
+    return trackPageDependencies(
+      result,
+      pageDependencies,
+      this.createPageDependency
+    )
+  }
+
+  getAllNodes(args, pageDependencies) {
+    const { type } = args || {}
+
+    let result
+    if (!type) {
+      result = this.nodeStore.getNodes()
+    } else {
+      const nodeTypeNames = toNodeTypeNames(this.schema, type)
+      const nodes = nodeTypeNames.reduce(
+        (acc, typeName) => acc.concat(this.nodeStore.getNodesByType(typeName)),
+        []
+      )
+      result = nodes.filter(Boolean)
+    }
+
+    return trackPageDependencies(
+      result,
+      pageDependencies,
+      this.createPageDependency
+    )
+  }
+
+  async runQuery(args, pageDependencies) {
+    const { query, firstOnly, type } = args || {}
+
+    // We don't support querying union types (yet?), because the combined types
```

Comment:
We should rethink not supporting union types in `runQuery` later, as part of refactoring resolving nodes and run-sift (e.g. we cannot call `getFields()` on a union type there). Note that `getAllNodes` etc. _do_ support both unions and interfaces.